framework/src/modules/interoperability/sidechain/commands/mainchain_registration.ts

```diff
@@ -12,25 +12,262 @@
+		const {
+			getAPIContext,
+			header,
+			getStore,
+			params: { ownChainID, ownName, mainchainValidators, aggregationBits, signature },
+		} = context;
+		const mainchainIdAsKey = getIDAsKeyForStore(MAINCHAIN_ID);
+		const { validators, certificateThreshold } = await this._bftAPI.getBFTParameters(
+			getAPIContext(),
+			header.height,
+		);
+		const activeValidators: ActiveValidators[] = [];
+		const keyList: Buffer[] = [];
+		const weights: bigint[] = [];
+		for (const v of validators) {
+			const { blsKey } = await this._validatorsAPI.getValidatorAccount(getAPIContext(), v.address);
+			activeValidators.push({
+				blsKey,
+				bftWeight: v.bftWeight,
+			});
+		}
+		sortValidatorsByBLSKey(activeValidators);
+		for (const v of activeValidators) {
+			keyList.push(v.blsKey);
+			weights.push(v.bftWeight);
+		}
+		const message = codec.encode(registrationSignatureMessageSchema, {
+			ownChainID,
+			ownName,
+			mainchainValidators,
+		});
+		verifyWeightedAggSig(
+			keyList,
+			aggregationBits,
+			signature,
+			TAG_CHAIN_REG_MESSAGE,
+			context.networkIdentifier,
+			message,
+			weights,
+			certificateThreshold,
+		);
+
+		const chainSubstore = getStore(MODULE_ID_INTEROPERABILITY, STORE_PREFIX_CHAIN_DATA);
+		await chainSubstore.setWithSchema(
+			mainchainIdAsKey,
+			{
+				name: MAINCHAIN_NAME,
+				networkID: MAINCHAIN_NETWORK_ID,
+				lastCertificate: {
+					height: 0,
+					timestamp: 0,
+					stateRoot: EMPTY_HASH,
+					validatorsHash: computeValidatorsHash(mainchainValidators, BigInt(THRESHOLD_MAINCHAIN)),
+				},
+				status: CHAIN_REGISTERED,
+			},
+			chainAccountSchema,
+		);
+
+		const channelSubstore = getStore(MODULE_ID_INTEROPERABILITY, STORE_PREFIX_CHANNEL_DATA);
+		await channelSubstore.setWithSchema(
+			mainchainIdAsKey,
+			{
+				inbox: { root: EMPTY_HASH, appendPath: [], size: 0 },
+				outbox: { root: EMPTY_HASH, appendPath: [], size: 0 },
+				partnerChainOutboxRoot: EMPTY_HASH,
+				messageFeeTokenID: { chainID: MAINCHAIN_ID, localID: 0 },
+			},
+			channelSchema,
+		);
+
+		const interoperabilityStore = this.getInteroperabilityStore(getStore);
+
+		const encodedParams = codec.encode(registrationCCMParamsSchema, {
+			networkID: MAINCHAIN_NETWORK_ID,
+			name: MAINCHAIN_NAME,
+			messageFeeTokenID: { chainID: MAINCHAIN_ID, localID: 0 },
+		});
+		const ccm = {
+			nonce: BigInt(0),
+			moduleID: MODULE_ID_INTEROPERABILITY,
+			crossChainCommandID: CROSS_CHAIN_COMMAND_ID_REGISTRATION,
+			sendingChainID: ownChainID,
+			receivingChainID: MAINCHAIN_ID,
+			fee: BigInt(0),
+			status: CCM_STATUS_OK,
+			params: encodedParams,
+		};
+
+		await interoperabilityStore.sendInternal({
+			moduleID: MODULE_ID_INTEROPERABILITY,
+			crossChainCommandID: CROSS_CHAIN_COMMAND_ID_REGISTRATION,
+			receivingChainID: MAINCHAIN_ID,
+			fee: BigInt(0),
+			status: CCM_STATUS_OK,
+			params: encodedParams,
+			timestamp: Date.now(),
```

Comment:
same as other PR, but module cannot use undeterministic code. If timestamp is needed, it should use block header timestamp.
https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#execution-4